[libc][c11] implement ctime

zimirza opened the pull request:
This is an implementation of `ctime` and includes `ctime_r`.

According to documentation, `ctime` and `ctime_r` are defined as the following:

```c
char *ctime(const time_t *timep);
char *ctime_r(const time_t *restrict timep, char buf[restrict 26]);
```

closes #86567

zimirza committed:
[libc][c11] implement ctime (#86567)

zimirza committed:
[libc][c11] implement ctime

remove buffer length

zimirza committed:
[libc][c11] implement ctime

fix: buffer length

zimirza committed:
[libc][c11] implement ctime

add buffer length

zimirza committed:
[libc][c11] implement ctime

fix: ctime implementation

zimirza committed:
[libc][c11] implement ctime

fix: tests

zimirza committed:
[libc][ctime] ctime implementation

fix: function declaration

zimirza committed:
[libc][c11] ctime implementation

format code with `clang-format`

zimirza committed:
[libc][c11] implement ctime

implement `ctime_r`

zimirza committed:
[libc][c11] implement ctime

fix: `ctime_r` header file

zimirza committed:
[libc][c11] implement ctime

add `ctime_r` to `cmake`

michaelrj-google reviewed (commented):
This is a good start, but I think we can simplify the code a bit. You'll also need to update the build system in a few places:

- [ ] add `ctime` and `ctime_r` to [`libc/config/linux/x86_64`](https://github.com/llvm/llvm-project/blob/main/libc/config/linux/x86_64/entrypoints.txt#L1004) as well as any other targets you want to enable these on.
- [ ] add `ctime` and `ctime_r` to the headergen files
  - [ ] for newheadergen you can run the "add_function" command described here: https://libc.llvm.org/dev/header_generation.html (the standard is `stdc` and the guard and attribute should each be `null`).
  - [ ] for old headergen you'll still have to add them manually in [`libc/spec/stdc.td`](https://github.com/llvm/llvm-project/blob/main/libc/spec/stdc.td#L1590)

Feel free to reach out if you have questions on how to do any of this, and thank you for the contribution!

zimirza committed:
[libc][c11] implement ctime

add `ctime` and `ctime_r` to build system

zimirza committed:
[libc][c11] implement ctime

refactor `ctime` and `ctime_r` to use `asctime`

zimirza committed:
[libc][c11] implement ctime

fix: `ctime_r`

zimirza committed:
[libc][c11] implement ctime

add `ctime` and `ctime_r` to `aarch64` and `riscv` architecture

zimirza commented on libc/src/time/time_utils.h:
```diff
@@ -145,6 +148,50 @@ LIBC_INLINE char *asctime(const struct tm *timeptr, char *buffer,
   return buffer;
 }
 
+LIBC_INLINE char *ctime(const time_t *t_ptr, char *buffer,
```
Thank you for the feedback. I have refactored this. I just need to do some testing before marking this pull request "ready for review".

zimirza committed:
[libc][c11] implement ctime

add `ctime` and `ctime_r` to more architectures

zimirza committed:
[libc][c11] implement ctime

- add symbols for `ctime` and `ctime_r`
- add `ctime` and `ctime_r` to clang formatted files list
- add `ctime` to env32-c test

zimirza committed:
[libc][c11] implement ctime

add header for `time_t`

zimirza committed:
[libc][c11] implement ctime

removed `ctime` from analysis test

zimirza committed:
[libc][c11] implement ctime

fix: `time_t` definition in `stdc.td`

zimirza committed:
[libc][c11] implement ctime

fix: removed `ctime_s`

zimirza committed:
[libc][c11] implement ctime

add `ctime` and `ctime_r` to unsafe functions check

michaelrj-google commented on libc/src/time/time_utils.h:
```diff
@@ -62,6 +62,9 @@ struct TimeConstants {
   static constexpr int ASCTIME_BUFFER_SIZE = 256;
   static constexpr int ASCTIME_MAX_BYTES = 26;
 
+  static constexpr int CTIME_BUFFER_SIZE = 256;
```
I think these aren't needed anymore

zimirza replied:
That is correct. I have removed these and just used `asctime` constants.

michaelrj-google commented on libc/src/time/ctime_r.cpp:
```diff
@@ -0,0 +1,22 @@
+using LIBC_NAMESPACE::time_utils::TimeConstants;
+
+LLVM_LIBC_FUNCTION(char *, ctime_r, (const time_t *t_ptr, char *buffer)) {
+  return time_utils::asctime(localtime(t_ptr), buffer, TimeConstants::CTIME_MAX_BYTES);
```
is `localtime` implemented?

zimirza replied:
`localtime` is available in `time.h`, or maybe I should implement this?

michaelrj-google replied:
it hasn't been implemented in LLVM-libc yet, so it needs to be implemented for this PR to work. 
Implementing `localtime` should be a separate patch, but it doesn't need to be complete to start with since we don't have time zones set up yet. For now it's okay if it's implemented to be the same as `gmtime`.

zimirza replied:
Sure, I have now just used `gmtime_internal`. I have created an issue for implementing `localtime`, https://github.com/llvm/llvm-project/issues/107597.

zimirza replied:
I created a function called `localtime` in `time_utils.h` that uses `gmtime_internal`.

zimirza committed:
[libc][c11] implement ctime

fix: tests

zimirza committed:
[libc][c11] implement ctime

use asctime constants and remove ctime constants

zimirza committed:
[libc][c11] implement ctime

use `TimeTTypePtr` and remove `StructTimeTPtr`

zimirza committed:
[libc][c11] implement ctime

remove `ctime` and `ctime_r` from `clang-tidy`

this will be done in https://github.com/llvm/llvm-project/issues/107445

zimirza committed:
[libc][c11] implement ctime

fix: tests

zimirza commented on libc/spec/stdc.td:
```diff
@@ -1600,6 +1602,19 @@ def StdC : StandardSpec<"stdc"> {
+          FunctionSpec<
+              "ctime",
+              RetValSpec<CharPtr>,
+              [ArgSpec<StructTimeTPtr>]
```
That is correct. I have now removed `StructTimeTPtr`.

zimirza committed:
[libc][c11] implement ctime

add error if argument is greater than 32bit int and fixed tests

zimirza committed:
undo clang-tidy

zimirza committed:
add header file for `time.h`

zimirza committed:
remove `time.h`

zimirza committed:
add function for `localtime` that temporarily use `gmtime`, since
`localtime` is not yet implemented

zimirza committed:
added github issue for implementing `localtime`

zimirza committed:
remove `struct tm` from `localtime`

lntue commented on libc/src/time/ctime.cpp:
```diff
@@ -0,0 +1,26 @@
+//===----------------------------------------------------------------------===//
+
+#include "ctime.h"
+#include "../__support/common.h"
+#include "../__support/macros/config.h"
```
Use path relative to the `libc/` folder:
```
#include "src/__support/common.h"
...
```

zimirza replied:
Thank you for the feedback. I have used relative path now.

lntue commented on libc/src/time/ctime.cpp:
```diff
@@ -0,0 +1,26 @@
+  return time_utils::asctime(localtime(t_ptr), buffer, TimeConstants::ASCTIME_MAX_BYTES);
+}
+
+} // namespace LIBC_NAMESPACE_DECL
```
Nit: missing new line at the end of file.

zimirza replied:
I have added newline at the end of all new files in this pull request.

zimirza committed:
use relative path

lntue commented on libc/src/time/time_utils.h:
```diff
@@ -156,6 +156,13 @@ LIBC_INLINE struct tm *gmtime_internal(const time_t *timer, struct tm *result) {
+// TODO: localtime is not yet implemented and a temporary solution is to
+//       use gmtime, https://github.com/llvm/llvm-project/issues/107597
+LIBC_INLINE struct tm *localtime(const time_t *timer, struct tm *result) {
+  struct tm *result = {0};
```
This one shadows the second input parameter?

zimirza replied:
`localtime` should only have one argument, so I removed `result`.

lntue commented on libc/test/src/time/ctime_r_test.cpp:
```diff
@@ -0,0 +1,63 @@
+  char* result = call_ctime(&t);
+  ASSERT_ERRNO_EQ(EINVAL);
+  ASSERT_STREQ(nullptr, result);
+}
```
Nit: Missing new line at the end of file.

zimirza committed:
add newline at end of files

zimirza committed:
add comment for tests

lntue commented on libc/src/time/ctime.h:
```diff
@@ -0,0 +1,20 @@
+
+} // namespace LIBC_NAMESPACE_DECL
+
+#endif // LLVM_LIBC_SRC_TIME_CTIME_H
```
Nit: missing new line at the end of file.

lntue commented on libc/src/time/ctime_r.cpp:
```diff
@@ -0,0 +1,25 @@
+  return time_utils::asctime(localtime(t_ptr), buffer, TimeConstants::ASCTIME_MAX_BYTES);
+}
+
+} // namespace LIBC_NAMESPACE_DECL
```
Nit: missing new line at the end of file.

lntue commented on libc/test/src/time/ctime_test.cpp:
```diff
@@ -0,0 +1,43 @@
+  char* result = call_ctime(&t);
+  ASSERT_ERRNO_EQ(EINVAL);
+  ASSERT_STREQ(nullptr, result);
+}
```
Nit: missing new line at the end of file.

lntue commented on libc/src/time/ctime_r.cpp:
```diff
@@ -0,0 +1,25 @@
+using LIBC_NAMESPACE::time_utils::TimeConstants;
+
+LLVM_LIBC_FUNCTION(char *, ctime_r, (const time_t *t_ptr, char *buffer)) {
+  if (t_ptr > 2147483647) {
```
Is there special named constant for this?

zimirza replied:
I have only found these references outside of tests:

```
./mlir/lib/Dialect/GPU/TransformOps/Utils.cpp:244:  static constexpr int maxTotalGriddim = 2147483647;
./mlir/lib/Dialect/GPU/TransformOps/Utils.cpp:245:  static constexpr int maxGriddimx = 2147483647;
```

It should really be in `time_utils.h`. I can add it to `TimeConstants` struct.

lntue commented on libc/src/time/ctime_r.cpp:
```diff
@@ -0,0 +1,25 @@
+
+LLVM_LIBC_FUNCTION(char *, ctime_r, (const time_t *t_ptr, char *buffer)) {
+  if (t_ptr > 2147483647) {
+    return llvm::createStringError(llvm::errc::invalid_argument, "ERROR");
```
I don't think `llvm::createStringError` and `llvm::errc::invalid_argument` are defined anywhere in `libc`, and we will not want the implementation to depend on other parts of llvm.

zimirza replied:
If `time_t` is greater than 32bit value, I now return `nullptr` here.

zimirza committed:
add newline to end of files

zimirza committed:
return `nullptr` if `time_t` is greater than 32bit value

zimirza committed:
return `nullptr` if `time_t` is greater than 32bit value

zimirza committed:
add 32bit maximum value to `TimeConstants` struct

lntue commented on libc/src/time/ctime_r.h:
```diff
@@ -0,0 +1,21 @@
+#define LLVM_LIBC_SRC_TIME_CTIME_R_H
+
+#include "src/__support/macros/config.h"
+#include <time.h>
```
Can you include the proxy header `"hdr/types/time_t.h"`, add `libc.hdr.types.time_t` to the target's cmake dependency, and remove `<time.h>` in here?

zimirza replied:
Sure, I removed `time.h` and added `hdr/types/time_t.h`, and updated cmake dependencies.

zimirza committed:
use `time_t.h` and include it in dependencies for `ctime` and `ctime_t`

zimirza committed:
update cmake for tests

zimirza committed:
remove symbols, since these will be implemented in
https://github.com/llvm/llvm-project/issues/107445

zimirza committed:
update cmake for tests

michaelrj-google commented on libc/src/time/ctime.cpp:
```diff
@@ -0,0 +1,26 @@
+    return nullptr;
+  }
+  static char buffer[TimeConstants::ASCTIME_BUFFER_SIZE];
+  return time_utils::asctime(localtime(t_ptr), buffer, TimeConstants::ASCTIME_MAX_BYTES);
```
make sure to call `time_utils::localtime` to avoid accidentally pulling in the system libc's version.

zimirza replied:
Yes, that is actually better. I added `time_utils::`.

michaelrj-google commented on libc/src/time/time_utils.h:
```diff
@@ -37,6 +37,7 @@ enum Month : int {
 };
 
 struct TimeConstants {
+  static constexpr int MAXIMUM_32_BIT_VALUE = 2147483647;
```
If you need the max value for a signed 32 bit number, I'd recommend using `cpp::numeric_limits<int32_t>::max()`

zimirza replied:
I have removed the constant and used this function.

michaelrj-google commented on libc/src/time/ctime.cpp:
```diff
@@ -0,0 +1,26 @@
+using LIBC_NAMESPACE::time_utils::TimeConstants;
+
+LLVM_LIBC_FUNCTION(char *, ctime, (const time_t *t_ptr)) {
+  if (t_ptr > TimeConstants::MAXIMUM_32_BIT_VALUE) {
```
this is comparing the pointer and not the value, did you mean to do that?

zimirza replied:
Yes, that is correct. The pointer should be dereferenced. Fixed it now.

michaelrj-google commented on libc/test/src/time/ctime_r_test.cpp:
```diff
@@ -0,0 +1,64 @@
+
+using LIBC_NAMESPACE::time_utils::TimeConstants;
+
+// make check-libc LIBC_TEST_TARGET=call_ctime_r VERBOSE=1
```
if you just want to run just this test you can run `make libc.test.src.time.ctime_r_test` and it should build and run it on its own.

zimirza replied:
This is interesting. I did not know that. I have updated the comments. Thank you.

zimirza committed:
use `time_utils::localtime`

zimirza committed:
use `cpp::numeric_limits<int32_t>::max()`

zimirza committed:
pointer should be dereferenced

zimirza committed:
update comment for tests

michaelrj-google commented on libc/test/src/time/ctime_r_test.cpp:
```diff
@@ -0,0 +1,64 @@
+using LIBC_NAMESPACE::time_utils::TimeConstants;
+
+// make libc.test.src.time.ctime_r_test
+static inline char *call_ctime_r(time_t *t, char *buffer) {
```
why does this function exist? It seems like it would be simpler to just call `LIBC_NAMESPACE::ctime_r` in the tests below.

zimirza replied:
I have now removed the helper functions in both tests.

michaelrj-google commented on libc/test/src/time/ctime_test.cpp:
```diff
@@ -0,0 +1,43 @@
+//
+//===----------------------------------------------------------------------===//
+
+// make libc.test.src.time.ctime_test
```
You should remove this comment before landing, as well as the one in the other test.

zimirza replied:
Yes, I have now removed the comments for testing in both test files.

michaelrj-google commented on libc/src/time/time_utils.h:
```diff
@@ -156,6 +156,13 @@ LIBC_INLINE struct tm *gmtime_internal(const time_t *timer, struct tm *result) {
+// TODO: localtime is not yet implemented and a temporary solution is to
+//       use gmtime, https://github.com/llvm/llvm-project/issues/107597
+LIBC_INLINE struct tm *localtime(const time_t *timer) {
+  struct tm *result = {0};
```
this should be `static struct tm result;` and then you should pass `&result` below. Since a pointer to this memory will be returned to the caller we need to make sure it's static.

zimirza replied:
Sure, I have now made `result` `static`. I just noticed that `tm` gets intialized in `update_from_seconds`, so there is no need to initialize `result` in this function.

zimirza committed:
remove helper functions and update comments

zimirza committed:
`result` should be `static`

zimirza committed:
remove comments

michaelrj-google reviewed (approved):
LGTM with a minor nit

zimirza commented on clang/docs/tools/clang-formatted-files.txt:
```diff
@@ -3062,6 +3062,10 @@ libc/src/time/asctime.cpp
 libc/src/time/asctime.h
 libc/src/time/asctime_r.cpp
 libc/src/time/asctime_r.h
+libc/src/time/ctime.cpp
+libc/src/time/ctime.h
+libc/src/time/ctime_r.cpp
+libc/src/time/ctime_r.h
```
I forgot to delete those. I had created another issue for that. I will delete them now.

zimirza committed:
removed `ctime` and `ctime_r` from clang fmt

zimirza committed:
format code with clang-format

michaelrj-google commented:
All good, review is a normal part of LLVM development. Now that this is approved, do you need me to merge it for you?

zimirza commented:
That would be great, since I cannot merge in LLVM.

zimirza committed:
fix: function argument

michaelrj-google commented on libc/src/time/ctime.h:
```diff
@@ -0,0 +1,20 @@
+#ifndef LLVM_LIBC_SRC_TIME_CTIME_H
+#define LLVM_LIBC_SRC_TIME_CTIME_H
+
+#include "src/__support/macros/config.h"
```
you need to add `#include "hdr/types/time_t.h"` here.

michaelrj-google commented on libc/src/time/ctime.cpp:
```diff
@@ -0,0 +1,27 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "ctime.h"
```
you should make these includes match `ctime_r.cpp`

michaelrj-google commented on libc/src/time/ctime_r.cpp:
```diff
@@ -0,0 +1,26 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/time/ctime_r.h"
```
you need to add `#include "src/__support/CPP/limits.h"` here and in `ctime.cpp`

michaelrj-google commented on libc/test/src/time/ctime_test.cpp:
```diff
@@ -0,0 +1,38 @@
+}
+
+TEST(LlvmLibcCtime, ValidUnixTimestamp0) {
+  struct time_t t = 0;
```
You need to remove `struct` from the `time_t` variables, since they aren't structs they are typedefs (here and in `ctime_r_test.cpp`.)
```suggestion
  time_t t = 0;
```

michaelrj-google commented on libc/test/src/time/ctime_r_test.cpp:
```diff
@@ -0,0 +1,57 @@
+  struct time_t t = 2147483647;
+  char *result;
+  // 2038-01-19 03:14:07. Test with a valid buffer size.
+  char *result = LIBC_NAMESPACE::ctime_r(t, buffer);
```
The `result` variable is already created above, so you should either remove the `char *` here or remove line 46.

michaelrj-google commented on libc/test/src/time/ctime_r_test.cpp:
```diff
@@ -0,0 +1,57 @@
+
+TEST(LlvmLibcCtimeR, InvalidArgument) {
+  struct time_t t = 2147483648;
+  char *result = call_ctime(&t);
```
remove `call_ctime` here, and make this a call to `ctime_r` which also needs a buffer.

michaelrj-google commented on libc/test/src/time/ctime_r_test.cpp:
```diff
@@ -0,0 +1,57 @@
+  struct time_t t = 0;
+  char *result;
+  // 1970-01-01 00:00:00. Test with a valid buffer size.
+  result = LIBC_NAMESPACE::ctime_r(t, buffer);
```
here and below, you need to pass a pointer to `t` to `ctime_r`
```suggestion
  result = LIBC_NAMESPACE::ctime_r(&t, buffer);
```

michaelrj-google commented on libc/test/src/time/ctime_r_test.cpp:
```diff
@@ -0,0 +1,57 @@
+  ASSERT_STREQ(nullptr, result);
+
+  time_t t;
+  result = LIBC_NAMESPACE::ctime_r(&tm_data, nullptr);
```
did you mean to pass `&t` instead of `&tm_data`?
```suggestion
  result = LIBC_NAMESPACE::ctime_r(&t, nullptr);
```

michaelrj-google commented on libc/test/src/time/ctime_test.cpp:
```diff
@@ -0,0 +1,38 @@
+
+TEST(LlvmLibcCtime, ValidUnixTimestamp0) {
+  struct time_t t = 0;
+  char *result = LIBC_NAMESPACE::ctime(t);
```
here and below, you need to pass a pointer to `t` to `ctime`
```suggestion
  char *result = LIBC_NAMESPACE::ctime(&t);
```

michaelrj-google reviewed (changes requested):
I ran the tests locally and ran into some issues, these need to be fixed before we can merge this.

zimirza committed:
fix: pointers

zimirza committed:
fix: header files

zimirza commented:
I am so sorry about this. I have now fixed them all.

zimirza committed:
fix: pointers

michaelrj-google commented:
it's fine, just make sure you run the tests to see what's working and what's broken. I usually just run `ninja check-libc` but if you're using make it should be `make check-libc`. There are a couple more places where things are broken, I'll add comments in a bit when I get a chance.

michaelrj-google commented:
also before I forget: If the tests aren't running for you it might be because these tests only run in full build mode: https://libc.llvm.org/full_host_build.html

When switching to full build mode you'll need to clean out your build folder because otherwise there are some confusing errors.

zimirza commented:
Sure, I will try to run the full build mode. I will let you know when this is ready.

zimirza committed:
fix: pointers

zimirza committed:
fix: pointers for tests

zimirza committed:
fix: pointers for tests

zimirza commented:
I have now tested `ctime` and `ctime_r` locally. It seems to work now.

zimirza commented:
I am testing locally and did find some issues. I will be fixing it now.

zimirza committed:
fix: tests

zimirza commented:
Tests for `ctime` and `ctime_r` are now good.

zimirza committed:
refactor `ctime` and `ctime_r`

michaelrj-google reviewed (approved):
Thanks for fixing this, it's all working on my machine now so I'll plan on landing it on Monday (since I don't want to land right before I get off work on Friday).

I really appreciate you working through this and getting it completed. I know it took a lot of effort but it's worth it at the end to have something high quality like this.

zimirza commented:
Thank you for the thorough review. I really do appreciate it. I have a question. Since this pull request is a new API feature implementation, should I also update documentation for `ctime` and `ctime_r`?

michaelrj-google commented:
You're welcome, and yes updating the docs for ctime and ctime_r would be good. I don't know if there's an existing page tracking the status of the time functions (it'd be in `libc/docs`). If there isn't then you can skip it. Adding a whole new page for the time functions should probably be a separate patch.

nickdesaulniers commented:
Looks like #109892 was a follow up to this PR.

zimirza commented:
That is correct. Unfortunately, I did not compare my implementation with glibc's `ctime`. So, I created a new issue for this. Hope it is ok.
